feat(cli): Replace `anchor verify` with a robust `solana-verify` wrapper

BretasArthur1 opened the pull request:
This PR modernizes Anchor's program verification by replacing the legacy, custom `anchor verify` implementation with a robust **"smart wrapper"** around the ecosystem-standard `solana-verify` tool. This change aligns Anchor with community best practices, reduces the maintenance burden, and provides a significantly more powerful and user-friendly experience.

### Problem

The previous `anchor verify` command used a custom, Docker-based verification process. This implementation had several drawbacks:
*   **Maintenance Burden:** It required ongoing maintenance to keep its dependencies and logic in sync with the rapidly evolving Solana toolchain.
*   **Divergence from Standard:** It was not aligned with `solana-verify`, which is the community-audited and widely adopted standard for verifiable builds.
*   **Limited Functionality:** It lacked the full feature set of the official `solana-verify` tool, limiting its usefulness for advanced verification scenarios.

### Solution

This PR addresses these issues by refactoring `anchor verify` and enhancing the Anchor Version Manager (`avm`) to fully integrate `solana-verify`. The solution is composed of three key improvements:

**1. `avm` Enhancements: Automatic `solana-verify` Installation**

The Anchor Version Manager (`avm`) is now responsible for managing the `solana-verify` dependency, which provides a seamless user experience:
*   **Automatic Installation:** When a user runs `avm install`, it now automatically installs the latest `solana-verify` tool directly from its official GitHub repository. This ensures users always have a compatible, up-to-date verification tool.
*   **Local Path Installation:** `avm` now includes a `--path` flag (`avm install --path .`), a powerful new feature for Anchor contributors to easily build and test their local changes.
*   **Seamless Integration:** The `avm` `anchor` wrapper now intelligently adds the `solana-verify` binary to the `PATH`, making it available to `anchor-cli` without requiring any manual user configuration.

**2. A New `anchor verify` Smart Wrapper**

Instead of re-implementing every flag, the new `anchor verify` command is a "smart wrapper"

*   **User-Friendly Flags:** It provides clear, documented flags for the most common use cases, such as `--repo-url` and `--current-dir`.
*   **Future-Proof Pass-Through:** Any arguments not explicitly recognized by `anchor verify` are passed directly to `solana-verify`. This is the core of the design and means that as `solana-verify` adds new features, Anchor developers can use them *immediately* without waiting for an `anchor-cli` update. For example: `anchor verify <ID> --repo-url <URL> -- --some-new-solana-verify-flag`.

**3. Robust Integration Test**

A new, focused integration test (`tests/anchor-cli-idl/verify_test.sh`) has been added to ensure the end-to-end functionality is correct. The test:
*   Checks for dependencies (Docker).
*   Installs the local `anchor-cli` and `solana-verify` using the newly enhanced `avm` command.
*   Verifies that the `anchor verify` smart wrapper can successfully verify a known, public program (Phoenix Protocol) on Mainnet.
*   Correctly handles the expected outcomes of running in a CI/test environment.

Closes: #3655 

Resources:
https://github.com/Ellipsis-Labs/solana-verifiable-build
https://solana.com/pt/developers/guides/advanced/verified-builds

BretasArthur1 committed:
feat: add support for tuple types in space calculation

This commit introduces the ability to handle tuple types in the space calculation logic. A new `TestTupleStruct` is added to validate various tuple configurations, including nested and option types, ensuring accurate space calculations through comprehensive unit tests.

BretasArthur1 committed:
clean up unnacessary comments

BretasArthur1 committed:
chore: fmt fix

BretasArthur1 committed:
fix: fixing test assert

BretasArthur1 committed:
fix: fixing test assert

BretasArthur1 committed:
Merge branch 'solana-foundation:master' into master

BretasArthur1 committed:
feat: replace anchor verify to use solana verify under the hood

BretasArthur1 committed:
feat: Enhance installation process with support for local paths and solana-verify integration

- Made `get_bin_dir_path` public to allow external access.
- Updated `InstallTarget` enum to include a `Path` variant for local repository installations.
- Modified `install_version` function to handle installation from a local path, including reading the version from the Cargo.toml manifest.
- Integrated installation of `solana-verify` during the installation process.
- Updated CLI commands to support the new path option for installation.
- Enhanced verification command to allow for repository URL or current directory options.

BretasArthur1 commented:
<img width="850" height="485" alt="Screenshot 2025-07-11 at 4 04 37 PM" src="https://github.com/user-attachments/assets/ae30ac59-9fd9-4dcc-9a20-70e53e1252da" />

BretasArthur1 committed:
Merge branch 'solana-foundation:master' into master

BretasArthur1 committed:
Refactor: Resolve avm binary collision by unifying proxy logic

Aursen commented on avm/src/lib.rs:
```diff
@@ -53,7 +53,30 @@ pub fn ensure_paths() {
+
+    // Create a symlink from `anchor` to `avm` so that the user can run `anchor`
+    // from the command line.
+    #[cfg(unix)]
```
What happen for Windows target?

BretasArthur1 replied:
yo, I actually forgot abt the windows users 😂

Aursen reviewed (commented):
Good improvement after all

BretasArthur1 committed:
feat: Add symlink creation for anchor.exe on Windows

BretasArthur1 committed:
fix: fmt

jacobcreech commented on avm/src/lib.rs:
```diff
@@ -280,6 +354,28 @@ pub fn install_version(
+            "install",
+            "solana-verify",
+            "--git",
+            "https://github.com/Ellipsis-Labs/solana-verifiable-build",
```
Can we force installation on a specific git commit?

Aursen commented on avm/src/lib.rs:
```diff
@@ -53,7 +53,45 @@ pub fn ensure_paths() {
+    // from the command line.
+    #[cfg(unix)]
+    {
+        let anchor_in_bin = bin_dir.join("anchor");
```
I think the symlink is wrong, if you do this, anchor command will launch avm instead of anchor-0.31.1 binary (or the one installed)

BretasArthur1 replied:
fe9c7830268cef7098c631484a85ba786479f943 check this

BretasArthur1 committed:
feat: Specify version for solana-verify installation

BretasArthur1 committed:
feat: Improve anchor command proxying and add testing script

- Updated the main function to enhance the logic for determining if the binary is named `anchor`.
- Introduced a new test script to verify that `anchor` commands correctly proxy to the installed Anchor CLI binary, ensuring expected behavior for both `avm` and `anchor` commands.

BretasArthur1 committed:
docs: Add caution note regarding mainnet wallet usage in verification test script

BretasArthur1 commented:
Sure

BretasArthur1 committed:
fix: CI fix

BretasArthur1 committed:
docs: add feature to the changelog

Aursen reviewed (approved):
On linux is working properly. On windows you need to have OpenSSL to build solana-verify it's a bit painful to get it but it's working.
So it's look good to me

Aursen committed:
chore: fix clippy complains (#3776)

* chore: fix clippy complains

* fix lints

* simplify some parts

---------

Co-authored-by: Aursen <aursen@users.noreply.github.com>

BretasArthur1 committed:
Merge branch 'master' into master

BretasArthur1 committed:
fix: correct median priority fee calculation in get_recommended_micro_lamport_fee function

BretasArthur1 committed:
fix: clippy
